marlowe-runtime/tx-api/Language/Marlowe/Runtime/Transaction/Api.hs

```diff
@@ -79,10 +83,80 @@ import Network.Protocol.ChainSeek.Types (SchemaVersion)
 import Network.Protocol.Handshake.Types (HasSignature(..))
 import Network.Protocol.Job.Types
 
+data CIP25MetadataFileDetails = CIP25MetadataFileDetails
+  { name :: Text
+  , mediaType :: Text
+  , src :: [Text]
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+
+instance Aeson.ToJSON CIP25MetadataFileDetails where
+  toJSON CIP25MetadataFileDetails {..} = Aeson.Object $
+      [ ("name", Aeson.String name)
+      , ("mediaType", Aeson.String mediaType)
+      ]
+        <> case src of
+          [x] -> [("src", Aeson.String x)]
+          xs  -> [("src", toJSON xs)]
+
+data CIP25MetadataDetails = CIP25MetadataDetails
+  { name :: Text
+  , image :: [Text]
+  , mediaType :: Maybe Text
+  , description :: [Text]
+  , files :: [CIP25MetadataFileDetails]
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+
+instance Aeson.ToJSON CIP25MetadataDetails where
+  toJSON CIP25MetadataDetails {..} =
+    Aeson.Object $
+      [ ("name", Aeson.String name) ]
+        <> case image of
+            [x] -> [("image", Aeson.String x)]
+            xs  -> [("image", toJSON xs)]
+        <> case mediaType of
+            Nothing -> []
+            Just x  -> [("mediatype", Aeson.String x)]
+        <> case description of
+            []  -> []
+            [x] -> [("description", Aeson.String x)]
+            xs  -> [("description", toJSON xs)]
+        <> case files of
+            [] -> []
+            xs -> [("files", toJSON xs)]
+
+data CIP25MetadataLabel = CIP25MetadataLabel
+  { policyIds :: Map PolicyId (Map TokenName CIP25MetadataDetails)
+  , version :: Maybe Int
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+
+instance Aeson.ToJSON CIP25MetadataLabel where
+  toJSON = \case
+    CIP25MetadataLabel {version = Nothing, policyIds} -> toJSON policyIds
+    CIP25MetadataLabel {version = Just 1, policyIds = toJSON -> Aeson.Object policyIds} ->
+      -- // TODO test this to make sure it's correct!
+      Aeson.Object (policyIds <> [("version", Aeson.Number 1)])
+    CIP25MetadataLabel {version = Just 2, policyIds = toJSON -> Aeson.Object policyIds} ->
+      -- // TODO test this to make sure it's correct!
+      Aeson.Object (policyIds <> [("version", Aeson.Number 2)])
+    CIP25MetadataLabel {version = Just n} -> error $ "Unknown version " <> show n
+
+-- // TODO test this to make sure it's correct!
+newtype CIP25Metadata = CIP25Metadata
+  { transactionMetadatumLabel :: CIP25MetadataLabel
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+
+instance Aeson.ToJSON CIP25Metadata where
+  toJSON CIP25Metadata {..} = Aeson.Object
+    [("721", toJSON transactionMetadatumLabel)]
```

Comment:
@jhbertra There's more work to be done but I think this shows my current approach to making [CIP-25](https://cips.cardano.org/cips/cip25/) compliant metadata.